deployment/terraform/utils.go

```diff
@@ -352,20 +355,116 @@ func getServerURL(output *Output, deploymentConfig *deployment.Config) string {
+// In that case, it also starts a goroutine that will refresh the credentials to
+// maintain them updated.
+func (t *Terraform) InitCreds() error {
 	regionOpt := awsconfig.WithRegion(t.config.AWSRegion)
+	var opts []func(*awsconfig.LoadOptions) error
+
+	opts = append(opts, regionOpt)
+
+	if t.config.AWSProfile != "" {
+		opts = append(opts, awsconfig.WithSharedConfigProfile(t.config.AWSProfile))
+	}
+
+	cfg, err := awsconfig.LoadDefaultConfig(context.Background(), opts...)
+	if err != nil {
+		return fmt.Errorf("unable to load default config: %w", err)
+	}
+
+	if t.config.AWSRoleARN != "" {
+		// See https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/credentials/stscreds#hdr-Assume_Role
+		// Create the credentials from AssumeRoleProvider to assume the role
+		// identified by AWSRoleARN, and use them in the loaded config
+		stsSvc := sts.NewFromConfig(cfg)
+		creds := stscreds.NewAssumeRoleProvider(stsSvc, t.config.AWSRoleARN)
+		cfg.Credentials = aws.NewCredentialsCache(creds)
+	}
+
+	t.awsCfg = cfg
+
+	if t.config.AWSRoleARN != "" {
+		// Refresh the credentials once so that the time of the interval below
+		// starts now
+		if err := t.refreshAWSCredentialsFromRole(); err != nil {
+			return fmt.Errorf("unable to refresh token: %w", err)
+		}
+
+		// Start a goroutine refreshing the credentials every AWSRoleTokenRefreshInterval.
+		// If it fails, exit the goroutine, since it needs the previous credentials to be
+		// valid for the refresh logic to work.
+		go func() {
+			ticker := time.Tick(AWSRoleTokenRefreshInterval)
+			for range ticker {
+				if err := t.refreshAWSCredentialsFromRole(); err != nil {
+					mlog.Error("unable to refresh token, stopping goroutine", mlog.Err(err))
+					return
+				}
+			}
+		}()
+	}
 
-	if t.config.AWSProfile == "" {
-		return awsconfig.LoadDefaultConfig(
-			context.Background(),
-			regionOpt,
-		)
+	return nil
+}
+
+// refreshAWSCredentialsFromRole refreshes the current credentials by assuming
+// the role again, with an expiry duration of one hour, which is the maximum
+// allowed when using role chaining
+func (t *Terraform) refreshAWSCredentialsFromRole() error {
+	t.awsCfgMut.Lock()
+	defer t.awsCfgMut.Unlock()
+
+	if t.config.AWSRoleARN == "" {
+		return fmt.Errorf("AWSRoleARN must be non-empty")
+	}
+
+	// Test current credentials before refresh: if they are not valid, we
+	// cannot refresh
+	if _, err := t.awsCfg.Credentials.Retrieve(context.Background()); err != nil {
+		return fmt.Errorf("failed to retrieve current credentials: %w", err)
+	}
+
+	// Create new STS service and assume role again
+	stsSvc := sts.NewFromConfig(t.awsCfg)
+	creds := stscreds.NewAssumeRoleProvider(stsSvc, t.config.AWSRoleARN, func(opt *stscreds.AssumeRoleOptions) { opt.Duration = AWSRoleTokenDuration })
+	newCredCache := aws.NewCredentialsCache(creds)
+
+	// Test new credentials
+	newCreds, err := newCredCache.Retrieve(context.Background())
```

Comment:
Could this (and related calls) block for a long time? (asking since we are not passing a timeout).

Reply:
Ah, great catch! Added a 30-second timeout for each call in the goroutine: https://github.com/mattermost/mattermost-load-test-ng/pull/940/commits/eb007124c0e9fdc66d047f54c4b6b82eba675b79